encoding/ccf/decode.go

```diff
@@ -0,0 +1,2039 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ccf
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+	goRuntime "runtime"
+
+	"github.com/fxamacker/cbor/v2"
+
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/runtime/common"
+	cadenceErrors "github.com/onflow/cadence/runtime/errors"
+)
+
+// CBORDecMode
+//
+// See https://github.com/fxamacker/cbor:
+// "For best performance, reuse EncMode and DecMode after creating them."
+var CBORDecMode = func() cbor.DecMode {
+	decMode, err := cbor.DecOptions{
+		IntDec:           cbor.IntDecConvertNone,
+		MaxArrayElements: math.MaxInt64,
+		MaxMapPairs:      math.MaxInt64,
+		MaxNestedLevels:  math.MaxInt16,
```

Comment:
I wonder if we should document this somehow. Theoretically the spec does not define such limits. Though we also want to limit the amount of decoded data to some sensible limit, so this is great.

Maybe we should add these limits also to the spec; or alternatively just document this for the codec, e.g. add a comment noting it on e.g. `Decoder`

Reply:
> Theoretically the spec does not define such limits. Though we also want to limit the amount of decoded data to some sensible limit, so this is great.

Yeah, that is exactly my thoughts too.  The CCF specs doesn't set these limits and it currently refers to CBOR RFC 8949 for security considerations (which mentions but doesn't specify CBOR decoder limits).

- 565f84e6576a6148f6b76d77cd2e5f68b752bd57 - Reduced max element limits for arrays and maps.
- 12b7116000c8f9f6900f009d775c6dfa636e8f6e - Reject indefinite length CBOR data items in CCF decoder (spotted this while rereading section 10 of RFC 8949).

There's an [issue](https://github.com/fxamacker/ccf_draft/issues/64) opened last month for CCF Specs to update it's section on Security Considerations, so I'll go ahead and do that.